Add SDKAnalysisLevel and UsingMicrosoftNETSdk to  packageSpec

Nigusu-Allehu opened the pull request:
<!-- DO NOT MODIFY OR DELETE THIS TEMPLATE. IT IS USED IN AUTOMATION. -->
## Bug

<!-- Search https://github.com/NuGet/Home/issues, and create one if you can't find a suitable issue. -->
<!-- Paste the full link, like https://github.com/nuget/home/issues/1000. GitHub will render is neatly. -->
Fixes: https://github.com/NuGet/Home/issues/13309

Regression? Last working version:

## Description
<!-- Add details about the fix. Include any information that would help the maintainer review this change effective. -->
* This PR follows https://github.com/dotnet/designs/pull/308, to add SDKAnalysisLevel to NuGet.
* The dotnet sdk installer sets the `SDKAnalysisLevel` value in `Microsoft.NETCoreSdk.BundledVersions.props` which is consumed early by msbuild. This is then consumed during restore.
* This pr allows NuGet to read `SdkAnalysisLevel` and `UsingMicrosoftNetSdk` and have them ready in package spec of a project.

### dotnet / msbuild 
* Since they use the RestoreTask task in NuGet.targets which calls `src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs` to get the dgspec content, update `MsbuildRestoreUtility` to get `SdkAnalysisLevel` and `UsingMicrosoftNETSdk ` properties.

### nuget.exe 
* Generates a temproary target file which will be used to generate a restore graph. 
  *  * This PR adds a property in the dgspec being generated called `SDKAnalysisLevel`  and `UsingMicrosoftNETSdk ` which  will then be consumed by nuget.exe restore command  and then passed to the restore runner similar to dotnet /msbuild restore.
  the dgspec file would looks as follows 
```
{
  "format": 1,
  "restore": {
    "C:\\demo\\hello\\hello\\hello.csproj": {}
  },
  "projects": {
    "C:\\demo\\hello\\hello\\hello.csproj": {
      "SDKAnalysisLevel": "9.0.100",
      "UsingMicrosoftNETSdk ": "true",
      "version": "1.0.0",
      "restore": {
        "projectUniqueName": "C:\\demo\\hello\\hello\\hello.csproj",
        "projectName": "hello",
        "projectPath": "C:\\demo\\hello\\hello\\hello.csproj",
        "packagesPath": "C:\\Nuget",
        "outputPath": "C:\\demo\\hello\\hello\\obj\\",
        "projectStyle": "PackageReference",
        "fallbackFolders": [
          "C:\\Program Files (x86)\\Microsoft Visual Studio\\Shared\\NuGetPackages"
        ],
        "configFilePaths": [
          "C:\\Users\\nyenework\\AppData\\Roaming\\NuGet\\NuGet.Config",
          "C:\\Program Files (x86)\\NuGet\\Config\\Microsoft.VisualStudio.FallbackLocation.config",
          "C:\\Program Files (x86)\\NuGet\\Config\\Microsoft.VisualStudio.Offline.config"
        ],
        "originalTargetFrameworks": [
          "net8.0"
        ],
        "sources": {
          "C:\\Program Files (x86)\\Microsoft SDKs\\NuGetPackages\\": {},
          "C:\\Program Files\\dotnet\\library-packs": {},
          "https://api.nuget.org/v3/index.json": {}
        },
        "frameworks": {
          "net8.0": {
            "targetAlias": "net8.0",
            "projectReferences": {}
          }
        },
        "warningProperties": {
          "warnAsError": [
            "NU1605"
          ]
        },
        "restoreAuditProperties": {
          "enableAudit": "true",
          "auditLevel": "low",
          "auditMode": "direct"
        }
      },
      "frameworks": {
        "net8.0": {
          "targetAlias": "net8.0",
          "dependencies": {
            "Newtonsoft.Json": {
              "target": "Package",
              "version": "[13.0.3, )"
            },
            "System.Text.Json": {
              "target": "Package",
              "version": "[8.0.3, )"
            }
          },
          "imports": [
            "net461",
            "net462",
            "net47",
            "net471",
            "net472",
            "net48",
            "net481"
          ],
          "assetTargetFallback": true,
          "warn": true,
          "frameworkReferences": {
            "Microsoft.NETCore.App": {
              "privateAssets": "all"
            }
          },
          "runtimeIdentifierGraphPath": "C:\\Program Files\\dotnet\\sdk\\8.0.301/PortableRuntimeIdentifierGraph.json"
        }
      }
    }
  }
}
```

### VS
* Get `SdkAnalysisLevel` and `UsingMicrosoftNETSdk ` values and add them to the corresponding project `packageSpec`

## PR Checklist

- [x] PR has a meaningful title
- [ ] PR has a linked issue.
- [ ] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  <!-- Describe why you haven't added automation. -->
  - [ ] Test exception
  - **OR**
  - [ ] N/A <!-- Infrastructure, documentation etc. -->

- **Documentation**
  <!-- Please link the PR/issue if appropriate -->https://github.com/NuGet/Home/issues/13603
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

Nigusu-Allehu committed:
Add SDKAnalysisLevel to RestoreTask

Nigusu-Allehu committed:
Cleanup

Nigusu-Allehu commented:
@NuGet/nuget-client  what do you think? Should we follow https://github.com/dotnet/installer/pull/18687 and add the default SDKAnalysisValues to NuGet manually for each release branch?

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets:
```diff
@@ -105,6 +105,11 @@ Copyright (c) .NET Foundation. All rights reserved.
 
+  <PropertyGroup>
+    <!--SDK Analysis level value should be updated in correspondance with the version of SDK this targeting-->
+    <SDKAnalysisLevel Condition=" '$(SDKAnalysisLevel)' == ''">9.1.00</SDKAnalysisLevel>
```
We are not supposed to set this.

Somewhere in the .NET SDK, this value will be set, and we should only consume it.

Nigusu-Allehu replied:
You are right. The dotnet sdk installer sets the value in `Microsoft.NETCoreSdk.BundledVersions.props` which is consumed early by msbuild.

Nigusu-Allehu committed:
Consume default value of SDKAnalysis from Microsoft.NETCoreSdk.BundledVersions

Nigusu-Allehu committed:
white space

zivkan commented on src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets:
```diff
@@ -179,7 +179,8 @@ Copyright (c) .NET Foundation. All rights reserved.
       RestorePackagesConfig="$(RestorePackagesConfig)"
-      EmbedFilesInBinlog="$(RestoreEmbedFilesInBinlog)">
+      EmbedFilesInBinlog="$(RestoreEmbedFilesInBinlog)"
+      SDKAnalysisLevel="$(SDKAnalysisLevel)">
```
Either this is not the right place to pass through the SDK analysis level, or we'll need a different design for Visual Studio and also NuGet.exe.

Within VS, we only get per-project information, there is no "whole solution" properties. In fact, Restore is quite unique, as I'm not aware of any other MSBuild task that runs at a solution level, everything else is per-project.  This also applies to `nuget.exe restore`, when we read PackageReference inputs.

Therefore, I'm fairly confident that we'll need to include this sdk analysis level in the PackageSpec, so that it's available to RestoreCommand, no matter which restore is being used ("normal" MSBuild restore, "static graph" restore, nuget.exe, VS).

zivkan commented on src/NuGet.Core/NuGet.Build.Tasks/RestoreTask.cs:
```diff
@@ -96,6 +96,8 @@ public class RestoreTask : Microsoft.Build.Utilities.Task, ICancelableTask, IDis
         /// </summary>
         public string EmbedFilesInBinlog { get; set; }
 
+        public string SDKAnalysisLevel { get; set; }
```
```suggestion
        public string SdkAnalysisLevel { get; set; }
```

It's convention to have initialisms and acronyms only upper case the first letter, which is why you see APIs like `XmlReader`, not `XMLReader`, `JsonDocument`, not `JSONDocument`, `SqlClient`, not `SQLClient`, etc.  I recently had a disagreement with Nikolche about MSBuild, because I believe if we follow convention, it should be `MsBuild`. But I think this is the only potential exception.

kartheekp-ms commented on src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets:
```diff
@@ -179,7 +179,8 @@ Copyright (c) .NET Foundation. All rights reserved.
       RestorePackagesConfig="$(RestorePackagesConfig)"
-      EmbedFilesInBinlog="$(RestoreEmbedFilesInBinlog)">
+      EmbedFilesInBinlog="$(RestoreEmbedFilesInBinlog)"
+      SDKAnalysisLevel="$(SDKAnalysisLevel)">
```
For my learning, does the `SDKAnalysisLevel` property require an older version of the SDK to be installed along with the newer version?

Nigusu-Allehu replied:
It does not. It can be used in one SDK version. Assume a user installed a newer SDK version but does not yet have a way to deal with the new warnings and errors that come with the new SDK. That is when they can use SDKAnalysisLevel to tell their new SDK they don't want these new warnings and errors by lowering the analysis level. As a result, whenever we introduce new errors and warnings, we should make sure they are considerate of the SDKAnalysisLevel

kartheekp-ms replied:
I don't fully understand why NuGet should have the SDKAnalysisLevel property ready for future consumption. Given that this is a new feature, I think we need a spec to fully understand the motivation behind this change. I have a few questions, such as why we need these properties in nuget.exe and what future requirements we have that necessitate this property. I was under the impression that we could use the new SDK analysis property similar to Andy's changes in the [commit](https://github.com/NuGet/NuGet.Client/commit/073a073d13b319f2d1d96af29d77f9dacd759053).

Nigusu-Allehu committed:
nuget.exe sdkanalysislevel impelementation

Nigusu-Allehu committed:
clean up

Nigusu-Allehu committed:
clean up

Nigusu-Allehu commented on src/NuGet.Core/NuGet.ProjectModel/PackageSpecWriter.cs:
```diff
@@ -65,6 +65,7 @@ internal static void Write(PackageSpec packageSpec, IObjectWriter writer, bool h
             SetDictionaryValues(writer, "scripts", packageSpec.Scripts);
 #pragma warning restore CS0612 // Type or member is obsolete
 
+            SetValue(writer, "SDKAnalysisLevel", packageSpec.SDKAnalysisLevel);
```
This adds `SDKAnalysisLevel` to the dgspec file that will be used by the restore command

Nigusu-Allehu commented on src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs:
```diff
@@ -211,6 +211,10 @@ internal static PackageSpec GetPackageSpec(JsonTextReader jsonReader, string nam
                         }
                         break;
+
+                    case "SDKAnalysisLevel":
```
reads `SDKAnalysisLevel`  value from the dgspec file so that restorecommand can pass it to restore runner

Nigusu-Allehu commented on src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs:
```diff
@@ -348,7 +348,31 @@ private static void AddRestoreSources(List<string> args, string[] sources)
+            var solutionPropsPath = Path.Combine(directoryOneLevelUp, "Directory.Solution.props");
+
             return GenerateMSBuildFile(
+                new XElement(MSBuildNamespace + "Import",
+                new XAttribute(XName.Get("Project"), solutionPropsPath),
+                new XAttribute(XName.Get("Condition"), $"Exists('{solutionPropsPath}')")),
+                new XElement(MSBuildNamespace + "Import",
+                new XAttribute(XName.Get("Project"), buildPropsPath),
+                new XAttribute(XName.Get("Condition"), $"Exists('{buildPropsPath}')")),
```
Imports `Directory.Build.props` and `Directory.Solution.prop` to the temporary MS build file that is generated in order to generate the dgspec file.

Nigusu-Allehu committed:
VS sdkAnalysisLevel

Nigusu-Allehu committed:
Get SdkAnalysis to restore args

Nigusu-Allehu committed:
remove unnecssary sdkanalysislevel

Nigusu-Allehu committed:
cleanup

Nigusu-Allehu committed:
add using dotnet sdk attribute

zivkan commented on src/NuGet.Clients/NuGet.CommandLine/Commands/RestoreCommand.cs:
```diff
@@ -199,6 +201,30 @@ public override async Task ExecuteCommandAsync()
+                }
+                else
+                {
+                    if (string.Compare(sdkAnalysisLevel, project.SdkAnalysisLevel, StringComparison.OrdinalIgnoreCase) < 0)
```
This is going to have unexpected results if some projects have SdkAnalysisLevel `10.0.100` and others have `9.0.100`.

zivkan commented on src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs:
```diff
@@ -348,7 +348,27 @@ private static void AddRestoreSources(List<string> args, string[] sources)
+            var directoryOneLevelUp = Path.GetDirectoryName(commonProjectDirectory);
+
+            // Paths for the Directory.Build.props one level up
+            var buildPropsPath = Path.Combine(directoryOneLevelUp, "Directory.Build.props");
```
You don't need to do this. When MSBuild is used to check each project in the solution, the project will import Directory.Build.props itself.

Nigusu-Allehu committed:
Add to sdkanalysislevel to packageSpec only

Nigusu-Allehu committed:
clean up

Nigusu-Allehu committed:
Write to dgspec file

Nigusu-Allehu committed:
rebase

Nigusu-Allehu committed:
make UsingSdk value nullable

Nigusu-Allehu commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs:
```diff
@@ -278,6 +278,8 @@ public static PackageSpec GetPackageSpec(IEnumerable<IMSBuildItem> items)
                     // NuGet audit properties
                     result.RestoreMetadata.RestoreAuditProperties = GetRestoreAuditProperties(specItem, GetAuditSuppressions(items));
+
+                    result.SdkAnalysisLevel = specItem.GetProperty("SdkAnalysisLevel");
```
Get `SdkAnalysisLevel` attribute from msbuild spec items

Nigusu-Allehu committed:
Clis and legacy projects

zivkan commented on src/NuGet.Core/NuGet.Build.Tasks/RestoreTask.cs:
```diff
@@ -159,6 +159,7 @@ private async Task<bool> ExecuteAsync(Common.ILogger log)
                 forceEvaluate: RestoreForceEvaluate,
                 hideWarningsAndErrors: HideWarningsAndErrors,
                 restorePC: RestorePackagesConfig,
+                cleanupAssetsForUnsupportedProjects: false,
```
was this an accident, perhaps some debugging code that was forgotten to be reverted?   It's not intuitive to me how it's related to this PR's title

Nigusu-Allehu replied:
I missed it when I was reverting previous edits

Nigusu-Allehu committed:
Read sdkanalysislevel for legacy

Nigusu-Allehu committed:
clean up

zivkan commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs:
```diff
@@ -308,6 +310,20 @@ public static PackageSpec GetPackageSpec(IEnumerable<IMSBuildItem> items)
+
+                if (!string.IsNullOrEmpty(isSdk))
+                {
+                    if (string.Equals(isSdk, "true", StringComparison.OrdinalIgnoreCase))
+                    {
+                        result.UsingMicrosoftNETSdk = true;
+                    }
+                    else if (string.Equals(isSdk, "false", StringComparison.OrdinalIgnoreCase))
+                    {
+                        result.UsingMicrosoftNETSdk = false;
+                    }
```
I didn't check the syntax, so this might not compile as is

```suggestion
                    if (bool.TryParse(isSdk, out bool value))
                    {
                        result.UsingMicrosoftNETSdk = value;
                    }
```

zivkan commented on src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs:
```diff
@@ -211,6 +211,25 @@ internal static PackageSpec GetPackageSpec(JsonTextReader jsonReader, string nam
+
+                        if (!string.IsNullOrEmpty(isSdk))
+                        {
+                            if (string.Equals(isSdk, "true", StringComparison.OrdinalIgnoreCase))
+                            {
+                                packageSpec.UsingMicrosoftNETSdk = true;
+                            }
+                            else if (string.Equals(isSdk, "false", StringComparison.OrdinalIgnoreCase))
+                            {
+                                packageSpec.UsingMicrosoftNETSdk = false;
+                            }
```
```suggestion
                            if (bool.TryParse(isSdk, out bool value))
                            {
                                packageSpec.UsingMicrosoftNETSdk = value;
                            }
```

zivkan commented on src/NuGet.Core/NuGet.ProjectModel/PackageSpec.cs:
```diff
@@ -162,6 +162,13 @@ public NuGetVersion Version
         /// <remarks>Optional. This is normally set for internal use only.</remarks>
         public ProjectRestoreMetadata RestoreMetadata { get; set; }
 
+        public string SdkAnalysisLevel { get; set; }
```
This is missing an XMLDoc comment

kartheekp-ms commented on src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets:
```diff
@@ -916,6 +916,8 @@ Copyright (c) .NET Foundation. All rights reserved.
         <NuGetAudit>$(NuGetAudit)</NuGetAudit>
         <NuGetAuditLevel>$(NuGetAuditLevel)</NuGetAuditLevel>
         <NuGetAuditMode>$(NuGetAuditMode)</NuGetAuditMode>
+        <SdkAnalysisLevel>$(SdkAnalysisLevel)</SdkAnalysisLevel>
```
I don't fully understand why NuGet should have the SDKAnalysisLevel property ready for future consumption. I have a few questions, such as why we need these properties in nuget.exe and what future requirements we have that necessitate this property. Given that this is a new feature, I think we need a spec to fully understand the motivation behind this change. I was under the impression that we could use the new SDK analysis property similar to Andy's changes in the [commit](https://github.com/NuGet/NuGet.Client/commit/073a073d13b319f2d1d96af29d77f9dacd759053). I am wondering if we are violating [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) principle.

zivkan replied:
Nigusu already linked the design spec in the PR description: https://github.com/dotnet/designs/pull/308

From NuGet's point of view, all new warnings (for example any future pack warnings, or any future restore warnings that aren't enabled only when a new feature is enabled), and any changes to defaults (like something that was previously a warning and we want to change to an error, or a feature that was previously opt-in and we want to change to opt-out) needs to be gated behind SdkAnalysisLevel

kartheekp-ms replied:
Thanks for the response, Andy. I noticed the design spec in the PR description and have already read it.

If we want to introduce a breaking change, we should hardcode the SDKAnalysisLevel property in the code as needed so that customers can opt in or out. I felt that your approach in [this commit](https://github.com/NuGet/NuGet.Client/commit/073a073d13b319f2d1d96af29d77f9dacd759053) is much cleaner and easier to maintain than hardcoding the SDK version across the codebase as needed. We always have some escape hatches for features via environment variables, NuGet.Config (for example, allowing insecure connections), Visual Studio feature flags, or NoWarns. Now, we are adding another option. Hence, I'm double-checking and thought that a spec would have helped clarify.

nkolev92 replied:
The property value being propagated is just implementation detail. 

Not every behavior change is a project specific configuration change. 

Say for example, we change when NU1605 is raised, say it's raised in some additional scenarios. 
That logic will never be part of the nuget.targets, so it's not fair to set that restriction.

Nigusu-Allehu committed:
static graph restore

zivkan commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs:
```diff
@@ -37,7 +37,7 @@ public sealed class LegacyPackageReferenceProject : PackageReferenceProject<Dict
         private readonly IVsProjectAdapter _vsProjectAdapter;
         private readonly IVsProjectThreadingService _threadingService;
-
+        private const bool IsSdk = false; // default value for non sdk projects
```
Personally, I don't see a big value in making this a const, when it's only used once.

zivkan commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs:
```diff
@@ -555,7 +555,9 @@ private async Task<PackageSpec> GetPackageSpecAsync(ISettings settings)
-                }
+                },
+                SdkAnalysisLevel = _vsProjectAdapter.BuildProperties.GetPropertyValue(ProjectBuildProperties.SdkAnalysisLevel),
+                UsingMicrosoftNETSdk = IsSdk,
```
So, a few weeks/months ago, we had a customer who is implementing their own custom project system in VS, use an SDK style project but NuGet used the project via `LegacyPackageReferenceProejct` because they weren't using CPS. Therefore, I think it's safer to get this property via the project's build properties service, rather than assuming it's false.

zivkan commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/VSNominationUtilities.cs:
```diff
@@ -712,5 +712,15 @@ private static string[] HandleClear(string[] input)
+
+        internal static bool GetUsingMicrosoftNETSdk(IReadOnlyList<IVsTargetFrameworkInfo4> targetFrameworks)
+        {
+            return string.Equals(GetSingleNonEvaluatedPropertyOrNull(targetFrameworks, nameof(ProjectBuildProperties.UsingMicrosoftNETSdk), v => v), "true", StringComparison.OrdinalIgnoreCase);
```
The `bool` type has `Parse` and `TryParse` methods. I'd use that, rather than string.Equals directly.

zivkan commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs:
```diff
@@ -385,7 +385,9 @@ internal static PackageSpec ToPackageSpec(ProjectNames projectNames, IVsProjectR
-                RestoreSettings = new ProjectRestoreSettings() { HideWarningsAndErrors = true }
+                RestoreSettings = new ProjectRestoreSettings() { HideWarningsAndErrors = true },
+                SdkAnalysisLevel = VSNominationUtilities.GetSdkAnalysisLevel(targetFrameworks),
+                UsingMicrosoftNETSdk = VSNominationUtilities.GetUsingMicrosoftNETSdk(targetFrameworks)
```
```suggestion
                UsingMicrosoftNETSdk = VSNominationUtilities.GetUsingMicrosoftNETSdk(targetFrameworks),
```

so any future properties only need to add lines, it won't need to modify an existing line

zivkan commented on src/NuGet.Core/NuGet.ProjectModel/PackageSpec.cs:
```diff
@@ -162,6 +162,18 @@ public NuGetVersion Version
+        /// When introducing a new warning or error use this property to
+        /// allow users to tell the sdk "treat me as if I were SDK x.y.z" and manage breaking changes
+        /// </summary>
+        public string SdkAnalysisLevel { get; set; }
```
As per the spec for SdkAnalysisLevel, it's going to be a version string and therefore it can't be compared using string comparisons. Since in the future we're going to compare this more and more, and therefore parsing it only once will have a perf benefit over needing to parse the string as a version multiple times.

```suggestion
        public NuGetVersion? SdkAnalysisLevel { get; set; }
```

zivkan commented on src/NuGet.Core/NuGet.ProjectModel/PackageSpecWriter.cs:
```diff
@@ -65,6 +65,12 @@ internal static void Write(PackageSpec packageSpec, IObjectWriter writer, bool h
+
+            if (packageSpec.UsingMicrosoftNETSdk is not null)
+            {
+                SetValue(writer, "UsingMicrosoftNETSdk", packageSpec.UsingMicrosoftNETSdk.ToString());
```
JSON supports bools, so write it as a bool rather than a string.

Also, given we default to false, we can write the value only when the value is true, so that parsing (reading) has a very minor perf benefit when reading the default value.

Although considering SDK style projects are more common and the only .NET project types that are "moving forward", perhaps we should default to true in the reader (meaning you need to handle the case when the property isn't defined in the json), and therefore only write to JSON when the value is false.

Nigusu-Allehu replied:
That is a good idea for performance. However, currently we default to null, unless we clearly get the information that UsingMicrosoftNetSdk is true or false.  Which i think is good for correctness.

zivkan replied:
Null doesn't have a semantic meaning. It means the same thing as false. So, I don't understand what you mean by correctness. When we use this property to make decisions about what warnings and errors to enable, null will have the same code path as false, always.

You could make it non-nullable in `PackageSpec.cs` and define the default value `public bool UsingMicrosoftNETSdk { get; set; } = true;`, so it will only be set to false if something explicit sets it to false, which we'd need to do everywhere we're doing `bool.TryParse` (if try returns false, set the value to false).

Nigusu-Allehu replied:
Defaulting to false when the `UsingMicrosoftNETSdk` value is absent might imply that we are certain about its value, even when it's actually undefined. I was wondering if there might be cases where the value isn't written to the dgSpec file. In such instances, should we automatically assume it's false instead of null, which typically indicates an unknown state?

zivkan commented on src/NuGet.Core/NuGet.ProjectModel/PackageSpec.cs:
You need to modify `GetHashCode` and `Equals` to handle the new properties, otherwise if customers change the value of either of these properties, then restore will incorrectly no-op and therefore the property value change won't affect restore unless customers do something else to force a restore.

zivkan reviewed (commented):
I didn't build a local copy of dotnet/project-system, but I tested all the other restore types and validated that after a restore the dgspec.json file in the obj file contained the values.

The version data type, and no-op checks are important improvements, but otherwise it looks good.

Nigusu-Allehu committed:
Address comments

Nigusu-Allehu committed:
white space

Nigusu-Allehu committed:
handle null value

zivkan commented on src/NuGet.Core/NuGet.ProjectModel/PackageSpec.cs:
```diff
@@ -193,6 +205,12 @@ public override int GetHashCode()
             hashCode.AddObject(RuntimeGraph);
             hashCode.AddObject(RestoreMetadata);
+            hashCode.AddObject(SdkAnalysisLevel);
+
+            if (UsingMicrosoftNETSdk != null)
+            {
+                hashCode.AddObject(UsingMicrosoftNETSdk.Value);
+            }
```
```suggestion
            hashCode.AddStruct(UsingMicrosoftNETSdk.Value);
```

zivkan reviewed (commented):
The bool in the package spec is still being read and written as a string, not as a bool.  Do a restore of a project with your branch, then open `obj\projectname.g.nuget.dgspec.json`, and you'll see a bunch of other properties with bool values (no quotes), compared to `UsingMicrosoftNETSdk` which will have quotes around the true/false string.

zivkan commented:
You also need to rebase your branch, to include the fixes (well, skipping) tests that are failing in CI.

Nigusu-Allehu committed:
merge coflict

Nigusu-Allehu committed:
white space

Nigusu-Allehu committed:
Use bool

zivkan commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs:
```diff
@@ -511,6 +511,20 @@ private async Task<PackageSpec> GetPackageSpecAsync(ISettings settings)
+
+            if (bool.TryParse(usingNetSdk, out bool result))
+            {
+                usingMicrosoftNetSdk = result;
```
My previous comments about making `UsingMicrosoftNetSdk` default to true applied to the dgspec/package spec file, not to "everything". This here is a logic bug. We know that non-SDK style projects won't set this property, so the value should be false when not set, but the way this is written now will incorrectly set it to true.

zivkan commented on src/NuGet.Core/NuGet.ProjectModel/PackageSpecWriter.cs:
```diff
@@ -65,6 +65,12 @@ internal static void Write(PackageSpec packageSpec, IObjectWriter writer, bool h
+                SetValue(writer, "SdkAnalysisLevel", packageSpec.SdkAnalysisLevel.ToString());
+            }
+
+            SetValueIfTrue(writer, "UsingMicrosoftNETSdk", packageSpec.UsingMicrosoftNETSdk);
```
My suggestion was to write the value to the json if the value is false, not if it's true. Since you took up my suggestion of setting the default value in `PackageSpec.cs` to true, and `JsonPackageSpecReader.cs` will only override the value when it's set in the json, this means this is a bug which will in effect mean that `UsingMicrosoftNETSdk` will always be set to true, never set the false.

zivkan commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs:
```diff
@@ -308,6 +315,13 @@ public static PackageSpec GetPackageSpec(IEnumerable<IMSBuildItem> items)
+                if (bool.TryParse(isSdk, out bool value))
+                {
+                    result.UsingMicrosoftNETSdk = value;
+                }
```
Since you took up my previous suggestion of making PackageSpec.cs default to true, that means this here now has a bug, because when the project file doesn't contain the property, the value is keeping its default `true` value, rather than being set to `false`.

```suggestion
                bool usingMicrosoftNETSdk = false;
                if (bool.TryParse(isSdk, out bool value))
                {
                    usingMicrosoftNETSdk = value;
                }
                result.UsingMicrosoftNETSdk = usingMicrosoftNETSdk;
```

zivkan commented on src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs:
```diff
@@ -211,6 +211,23 @@ internal static PackageSpec GetPackageSpec(JsonTextReader jsonReader, string nam
+                        }
+                        break;
+                    case "UsingMicrosoftNETSdk":
+                        string isSdk = jsonReader.ReadNextTokenAsString();
```
jsonReader has a specific API to read the value as a boolean.

`ReadNextTokenAsString` will cause a memory allocation, and `JsonPackageSpecReader` is one of our most performance sensitive parts of the code (everything related to restore is), so we must optimize for performance.

System.Text.Json is capable of parsing bool values without any memory allocations, so bool values must be read as bool, not allocated as strings and then parsed as a bool.

zivkan commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/VSNominationUtilities.cs:
```diff
@@ -712,5 +712,29 @@ private static string[] HandleClear(string[] input)
+            return sdkAnalysisLevel;
+        }
+
+        internal static bool? GetUsingMicrosoftNETSdk(IReadOnlyList<IVsTargetFrameworkInfo4> targetFrameworks)
+        {
+            string? usingNetSdk = GetSingleNonEvaluatedPropertyOrNull(targetFrameworks, nameof(ProjectBuildProperties.UsingMicrosoftNETSdk), v => v);
+
+            if (bool.TryParse(usingNetSdk, out var result))
+            {
+                return result;
+            }
+
+            return null;
```
This no longer compiles since you changed `UsingMicrosoftNETSdk` in `PackageSpec.cs` to be a `bool` rather than `bool?`.

Remember now that `PackageSpec.cs` defaults to `true`, that if the parsing fails, or if the string isn't set in the project, that this code needs to explicitly set the PackageSpec's property to false, do not leave the default.

zivkan commented on src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs:
```diff
@@ -810,8 +810,19 @@ private PackageSpec GetPackageSpec(IMSBuildProject project, IReadOnlyDictionary<
+
+            if (bool.TryParse(project.GetProperty("UsingMicrosoftNETSdk"), out bool value))
+            {
+                packageSpec.UsingMicrosoftNETSdk = value;
```
Like other comments, with `PackageSpec.cs` now defaulting this property value to `true` (so that `JsonPackageSpecReader.cs` doesn't need to handle the default == true value), this has a bug when the project doesn't set the property to any value, and we need to treat it as false.

zivkan commented on src/NuGet.Core/NuGet.ProjectModel/PackageSpec.cs:
```diff
@@ -162,6 +162,18 @@ public NuGetVersion Version
+        /// <summary>
+        /// Indicates that Microsoft.NET.Sdk is being used.
+        /// </summary>
+        public bool UsingMicrosoftNETSdk { get; set; } = true;
```
I've commented on several files how when we're reading the property from the project, we need to handle the property not being set as `false`, and explicitly set this value in that case.

An alternative is to let this bool keep its `false` default, and instead in `JsonPackageSpecReader.cs` to set the property value to `true` when the json doesn't contain the property. That way it only needs to be done in one place, not in every place we convert a project into a PackageSpec.

nkolev92 replied:
How are you going to use UsingMicrosoftNETSdk? 
Why in particular do we need it?

zivkan replied:
It'll be used to tell the difference between an SDK style project using an older version of the .NET SDK before they added SdkAnalysisLevel, and non-SDK style (legacy) project files, where we can make different decisions about default values.

nkolev92 replied:
What would those defaults be?

I'm trying to truth table that:

| SdkAnalysisLevel | UsingMicrosoftNETSdk | Explanation | 
| AnyValue | true | Happy case. Respect SDKAnalysisLevel |
| | true | ??? |
| | false | ??? |
| Any Value | false | ??? |

zivkan replied:
For NuGetAudit related changes, it would be:

|SdkAnalysisLevel|UsingMicrosoftNETSdk|Behaviour|
|--|--|--|
|Set|N/A| Set feature defaults per SdkAnalysisLevel|
|Not set|`true`|Assume SdkAnalysisLevel 8.0|
|Not set|`false` or not set|Assume the highest SdkAnalysisLevel each feature changes defaults for|

zivkan reviewed (commented):
I know you're under time pressure, so I'll leave it up to other people whether you merge this PR without writing tests, but I think the number of comments I've posted over the last week, and the bugs that were introduced in the last few days, is an example of why adding tests is important.

Implementing a feature incrementally via multiple pull requests does not mean "don't write any tests until the end", you can and should still write tests for the code you're changing in a PR, but it allows you to get early feedback, rather than implementing the entire feature, including tests, before getting any feedback from the rest of the team.

While I'm not sure any meaningful tests can be written for LegacyPackageReferenceProject,  your PR modifies PackageSpec.cs, and there exists a [PackageSpecTests.cs](https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.Tests/NuGet.ProjectModel.Test/PackageSpecTests.cs), which if you looked at might have helped you catch the `GetHashCode` and `Equals` method bugs sooner. And at some point we need those tests to reduce risk of regressions in the future.

Similarly, there are [PackageSpecWriterTests.cs](https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.Tests/NuGet.ProjectModel.Test/PackageSpecWriterTests.cs) and [JsonPackageSpecReaderTests.cs](https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.Tests/NuGet.ProjectModel.Test/JsonPackageSpecReaderTests.cs) and if tests were added to those for the new properties, perhaps you would have addressed issues before commiting the code that led to my feedback last week, and therefore this PR might have been able to be merged sooner, even though more up-front effort would have been needed to write those tests.

Nigusu-Allehu committed:
ci build

nkolev92 commented on src/NuGet.Core/NuGet.ProjectModel/PackageSpec.cs:
```diff
@@ -162,6 +162,18 @@ public NuGetVersion Version
+        /// When introducing a new warning or error use this property to
+        /// allow users to tell the sdk "treat me as if I were SDK x.y.z" and manage breaking changes
+        /// </summary>
+        public NuGetVersion SdkAnalysisLevel { get; set; }
```
I don't think this is a top level thing. 

https://github.com/NuGet/NuGet.Client/blob/20f05435be385abfe74737b6433dc80fd3b3b504/src/NuGet.Core/NuGet.ProjectModel/ProjectRestoreMetadata.cs should contain this instead. 

You'll see other unique values being represented there.

nkolev92 reviewed (changes requested):
I think we need to have this at a different location.

Nigusu-Allehu committed:
Make usre usingSDK value passed correctly

Nigusu-Allehu committed:
empty sdk analysis

Nigusu-Allehu committed:
read dgspec and default true

Nigusu-Allehu committed:
move sdkAnalysisLevel and UsingMicrosoftNetSdk to restoremetadata

Nigusu-Allehu committed:
address null metadata

Nigusu-Allehu commented:
I am working on adding the tests. In the mean time please feel free to review the implementation. I have moved the properties to RestoreMetadata or packagespec.

Nigusu-Allehu committed:
Read values from lock file

zivkan commented on src/NuGet.Core/NuGet.ProjectModel/Utf8JsonStreamReader.cs:
```diff
@@ -220,6 +220,17 @@ internal bool ReadNextTokenAsBoolOrFalse()
+        {
+            ThrowExceptionIfDisposed();
+
+            if (Read() && (TokenType == JsonTokenType.False || TokenType == JsonTokenType.True))
+            {
+                return GetBoolean();
+            }
+            return true;
```
If the JSON contains an invalid value (like a number, or some other random string, instead of a bool value), shouldn't we throw an exception, rather than silently ignoring the value?

What happens if other parts of the json are invalid, do we ignore it, or throw?

Nigusu-Allehu replied:
We ignore in other parts of the json: https://github.com/NuGet/NuGet.Client/blob/d01c264491805fef8d878816c2c7e338bf0d7290/src/NuGet.Core/NuGet.ProjectModel/Utf8JsonStreamReader.cs#L212
But I think we should of course throw an exception. The exception would probably not be thrown that often, as we are the ones generating the lock file

zivkan commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/VSNominationUtilities.cs:
```diff
@@ -712,5 +712,30 @@ private static string[] HandleClear(string[] input)
+
+            if (!string.IsNullOrEmpty(skdAnalysisLevelString))
+            {
+                sdkAnalysisLevel = new NuGetVersion(skdAnalysisLevelString);
```
If `SdkAnalysisLevel` can't be parsed as a version, the error experience should explicitly tell the customer to fix the value of `SdkAnalysisLevel`, hence the property name needs to be in any error message.

zivkan commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs:
```diff
@@ -511,6 +511,20 @@ private async Task<PackageSpec> GetPackageSpecAsync(ISettings settings)
+
+            if (skdAnalysisLevelString != null)
+            {
+                sdkAnalysisLevel = new NuGetVersion(skdAnalysisLevelString);
```
If SdkAnalysisLevel can't be parsed as a version, the error experience should explicitly tell the customer to fix the value of SdkAnalysisLevel, hence the property name needs to be in any error message.

zivkan commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs:
```diff
@@ -511,6 +511,20 @@ private async Task<PackageSpec> GetPackageSpecAsync(ISettings settings)
+            if (bool.TryParse(usingNetSdk, out bool result))
+            {
+                usingMicrosoftNetSdk = result;
+            };
```
What's the error experience when other boolean properties can't be parsed?  Do we silently ignore and use a default, do we warn and use a default, or do we error?

zivkan commented on src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs:
```diff
@@ -893,6 +893,23 @@ private PackageSpec GetPackageSpec(IMSBuildProject project, IReadOnlyDictionary<
+
+            if (skdAnalysisLevelString != null)
+            {
+                restoreMetadata.SdkAnalysisLevel = new NuGetVersion(skdAnalysisLevelString);
```
If SdkAnalysisLevel can't be parsed as a version, the error experience should explicitly tell the customer to fix the value of SdkAnalysisLevel, hence the property name needs to be in any error message.

zivkan commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs:
```diff
@@ -278,6 +278,13 @@ public static PackageSpec GetPackageSpec(IEnumerable<IMSBuildItem> items)
+
+                    if (skdAnalysisLevelString != null)
+                    {
+                        result.RestoreMetadata.SdkAnalysisLevel = new NuGetVersion(skdAnalysisLevelString);
```
If SdkAnalysisLevel can't be parsed as a version, the error experience should explicitly tell the customer to fix the value of SdkAnalysisLevel, hence the property name needs to be in any error message.

zivkan commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs:
```diff
@@ -308,6 +315,18 @@ public static PackageSpec GetPackageSpec(IEnumerable<IMSBuildItem> items)
+                }
+                else
+                {
+                    result.RestoreMetadata.UsingMicrosoftNETSdk = value;
+                    result.RestoreMetadata.UsingMicrosoftNETSdk = false;
```
```suggestion
                    result.RestoreMetadata.UsingMicrosoftNETSdk = false;
```

zivkan commented on src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.Utf8JsonStreamReader.cs:
```diff
@@ -225,6 +227,29 @@ internal static PackageSpec GetPackageSpec(ref Utf8JsonStreamReader jsonReader,
+                            packageSpec.RestoreMetadata = new ProjectRestoreMetadata();
+                        }
+
+                        packageSpec.RestoreMetadata.UsingMicrosoftNETSdk = jsonReader.ReadNextTokenAsBoolOrTrue();
```
As commented elsewhere, what's the error experience when other boolean properties have values, but can't be parsed as a bool?  Silently ignoring and using a default feels "wrong" to me. Maybe that's the convention for reading project properties, but this code is in the PackageSpec reader, and we own the package spec writer, so we can ensure that we never write an invalid value, and therefore we should never expect an invalid value.  Therefore, if this does have an invalid value, it means something is really messed up, so I think an exception should be thrown.

zivkan commented on src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.Utf8JsonStreamReader.cs:
```diff
@@ -225,6 +227,29 @@ internal static PackageSpec GetPackageSpec(ref Utf8JsonStreamReader jsonReader,
                             }
                         }
                     }
+                    else if (jsonReader.ValueTextEquals(UsingMicrosoftNETSdk))
```
This code is reading the root `PackageSpec` json, but that's not where these properties live in the updated PR. Now that it's part of `RestoreMetadata`, you need to move these functions to the appropriate place.

If you write unit tests, this mistake will become very apparent.

zivkan commented on src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs:
```diff
@@ -211,6 +211,19 @@ internal static PackageSpec GetPackageSpec(JsonTextReader jsonReader, string nam
                         }
                         break;
+
+                    case "SdkAnalysisLevel":
```
Similar to the other json spec reader, this code appears to be in the wrong place. There will never be an `SdkAnalysisLevel` property on the "root" `PackageSpec` object, because it's moved to `RestoreMetadata`.

zivkan reviewed (changes requested):
I found a few mistakes that would have been caught before pushing to github if tests were written. So, I've stopped reviewing the rest of the PR, because I feel like it's not ready for review yet.

Nigusu-Allehu committed:
test packagespec and packageSpecWriter

Nigusu-Allehu committed:
GetPackageSpec test

Nigusu-Allehu committed:
throw excpetion when reading invalid project properties

Nigusu-Allehu committed:
cealnup

Nigusu-Allehu committed:
Throw exception when parsing sdkAnalysisLevel from dgspec file

Nigusu-Allehu committed:
add tests

zivkan commented on src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs:
```diff
@@ -893,6 +893,38 @@ private PackageSpec GetPackageSpec(IMSBuildProject project, IReadOnlyDictionary<
+
+            string skdAnalysisLevelString = project.GetProperty("SdkAnalysisLevel");
+
+            if (skdAnalysisLevelString != null)
```
Did you test this?  MSBuild APIs generally treat undefined properties as empty strings, rather than nulls, so I would have thought that `!string.IsNullOrEmpty(sdkAnalysisLevelString)` would be needed. In any case, even if static graph APIs do use null rather than empty string, it might still be worthwhile using `IsNullorEmpty` just to be safe.

Same goes for `isSdk` below.

Nigusu-Allehu replied:
Yes, from my test when the value is not defined, GetProperty returns null as a result if somehow `sdkAnalysisLevelString` is an empty string that would something has gone wrong, and we would throw an exception when we try new NuGetVersion(""). However, we can also just ignore empty strings and nulls as you suggested.

GetProperty : https://github.com/NuGet/NuGet.Client/blob/6c642c2d63717acd4bf92050a42691928020eb89/src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildItemBase.cs#L29

zivkan commented on src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.Utf8JsonStreamReader.cs:
```diff
@@ -1169,6 +1173,32 @@ private static void ReadMSBuildMetadata(ref Utf8JsonStreamReader jsonReader, Pac
                     }
+                    else if (jsonReader.ValueTextEquals(UsingMicrosoftNETSdk))
+                    {
+                        usingMicrosoftNetSdk = jsonReader.ReadNextTokenAsBoolOrTrue(Encoding.UTF8.GetString(UsingMicrosoftNETSdk));
```
`Encoding.UTF8.GetString` will cause a heap allocation, and restore is our most performance critical code, so we must eliminate all unnecessary allocations.

One option is to pass the `byte[]` or `Span<byte>` to the `ReadNextTokenAsBoolOrTrue`, and only convert it to a string when the string is needed (in the error handling code), so it's not allocated in the non-error code path.

However, knowing a bit more about how .NET works, you can create a const string, which you pass to `Encoding.UTF8.GetBytes` where the `static readonly byte[]` is defined, because that `byte[]` is calculated at runtime, which means that if you inspect the compiled assembly, there's already a `"UsingMicrosoftNETSdk"` const string. Therefore, creating another const and using it here, instead of `UTF8.GetString` will avoid the allocation, and it won't result in any new additional string constants in the assembly.  However, since the variable name is identical, you can also use `nameof(UsingMicrosoftNETSdk)`.

So, to reiterate, anything related to restore, make sure you keep high performance code in mind, and avoid unnecessary heap allocations. There are several ways to achieve it, so you have flexibility on how you do it.

zivkan commented on src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs:
```diff
@@ -1154,6 +1156,40 @@ private static void ReadMSBuildMetadata(JsonTextReader jsonReader, PackageSpec p
+                        break;
+
+                    case "UsingMicrosoftNETSdk":
+                        string usingMicrosoftNetSdkString = jsonReader.ReadNextTokenAsString();
```
Newtonsoft.Json has a `ReadAsBool()` API: https://www.newtonsoft.com/json/help/html/M_Newtonsoft_Json_JsonReader_ReadAsBoolean.htm

Can it be used?

zivkan reviewed (commented):
Just one or two minor issues, then I'll be happy to sign off.  Thanks for all the effort!

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs:
```diff
@@ -511,6 +511,34 @@ private async Task<PackageSpec> GetPackageSpecAsync(ISettings settings)
             // Do not add new properties here. Use BuildProperties.GetPropertyValue instead, without DTE fallback.
 #pragma warning restore CS0618 // Type or member is obsolete
+            string skdAnalysisLevelString = _vsProjectAdapter.BuildProperties.GetPropertyValue(ProjectBuildProperties.SdkAnalysisLevel);
+            string usingNetSdk = _vsProjectAdapter.BuildProperties.GetPropertyValue(ProjectBuildProperties.UsingMicrosoftNETSdk);
```
Here you're in non-SDK based projects. 
So this should never been true.

I don't think you need the usingNetSdk codepaths.

zivkan replied:
earlier this year we had some external customer implementing their own VS project system hit NuGet's LegacyPackageReferenceProject code path, even though it was a net8.0 project.  I'd rather not put artificial limitations in NuGet and get support questions like that when it's so easy to just read the project property.

nkolev92 replied:
Did that work? I don't expect it'd work properly. Maybe in some cases in those, but it won't in general.

nkolev92 commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/VSNominationUtilities.cs:
```diff
@@ -712,5 +712,44 @@ private static string[] HandleClear(string[] input)
+
+        internal static NuGetVersion? GetSdkAnalysisLevel(IReadOnlyList<IVsTargetFrameworkInfo4> targetFrameworks)
+        {
+            string skdAnalysisLevelString = GetSingleNonEvaluatedPropertyOrNull(targetFrameworks, nameof(ProjectBuildProperties.SdkAnalysisLevel), v => v) ?? string.Empty;
```
Is SdkAnalysisLevel supposed to be per framework or global? 
https://github.com/dotnet/designs/blob/main/proposed/sdk-analysis-level.md doesn't clarify that. 

@baronfel any thoughts?

baronfel replied:
It's global, not per-TFM.

nkolev92 commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/VSNominationUtilities.cs:
```diff
@@ -712,5 +712,44 @@ private static string[] HandleClear(string[] input)
+            {
+                try
+                {
+                    sdkAnalysisLevel = new NuGetVersion(skdAnalysisLevelString);
```
nit: Just a pointer, there's a TryParse version which you can use to avoid the exception throwing/catcihng.

Nigusu-Allehu replied:
You are right. However, we would like to make sure the user is aware why their SdkAnalysisLevel value is not being used if they used an invalid value.

nkolev92 replied:
TryParse returns true/false depending if the value is parsable or not, so you can still report it.

nkolev92 replied:
Looking at it again, TryParse never throws, but Parse just calls TryParse and then throws. We catch that exception and then we rethrow it. 

Exceptions aren't free so best avoided.

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs:
```diff
@@ -893,6 +893,38 @@ private PackageSpec GetPackageSpec(IMSBuildProject project, IReadOnlyDictionary<
+            {
+                try
+                {
+                    restoreMetadata.SdkAnalysisLevel = new NuGetVersion(skdAnalysisLevelString);
```
Can you extract these in testable methods?
There's a precedent in this method itself, for example: WarningProperties.GetWarningProperties.

nkolev92 commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs:
```diff
@@ -308,6 +308,36 @@ public static PackageSpec GetPackageSpec(IEnumerable<IMSBuildItem> items)
                 result.RestoreMetadata.CentralPackageFloatingVersionsEnabled = isCentralPackageFloatingVersionsEnabled;
                 result.RestoreMetadata.CentralPackageTransitivePinningEnabled = isCentralPackageTransitivePinningEnabled;
+
+                string isSdk = specItem.GetProperty("UsingMicrosoftNETSdk");
```
Same as the other comment.
You should have a helper method that both this and static graph implementations call.

nkolev92 commented on src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs:
```diff
@@ -1154,6 +1156,40 @@ private static void ReadMSBuildMetadata(JsonTextReader jsonReader, PackageSpec p
+                            }
+                            catch (ArgumentException ex)
+                            {
+                                throw new ArgumentException(string.Format(CultureInfo.CurrentCulture, Strings.Invalid_AttributeValue, "SdkAnalysisLevel", skdAnalysisLevelString, "9.0.100"), ex);
```
This error handling is probably excessive since this is an internal type. 

I'd just use version parse, like the rest of the code here does.

Nigusu-Allehu replied:
That's how I had it implemented at first, but after @zivkan 's comments I added the custom exceptions. I agree it could be excessive but in the case somehow a dgspec file is edited and it contains an invalid value for these attributes, we will get an informative error telling the user which property exactly is causing the failure. The version parser would produce a general error on the other hand which might not be as helpful

nkolev92 commented on src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs:
```diff
@@ -1154,6 +1156,40 @@ private static void ReadMSBuildMetadata(JsonTextReader jsonReader, PackageSpec p
+                        }
+                        else
+                        {
+                            throw new ArgumentException(
```
Same here, there's precedent in this class. Makes the code easier to read.

nkolev92 commented on src/NuGet.Core/NuGet.ProjectModel/PackageSpecWriter.cs:
```diff
@@ -188,6 +187,11 @@ private static void SetMSBuildMetadata(IObjectWriter writer, PackageSpec package
 
+            if (packageSpec.RestoreMetadata.SdkAnalysisLevel is not null)
+            {
+                SetValue(writer, "SdkAnalysisLevel", packageSpec.RestoreMetadata.SdkAnalysisLevel.ToString());
```
SetValue has a null or empty check, just use that.

Nigusu-Allehu replied:
packageSpec.RestoreMetadata.SdkAnalysisLevel is a NugetVersion type, and it could be null. We are doing the null check here to prevent null reference when we do .ToString() before calling SetValue

nkolev92 replied:
Oh yeah great point. I missed that when reviewing. I assumed it was a string.

nkolev92 commented on src/NuGet.Core/NuGet.ProjectModel/PackageSpecWriter.cs:
```diff
@@ -210,6 +214,7 @@ private static void WriteMetadataBooleans(IObjectWriter writer, ProjectRestoreMe
             SetValueIfTrue(writer, "centralPackageFloatingVersionsEnabled", msbuildMetadata.CentralPackageFloatingVersionsEnabled);
             SetValueIfTrue(writer, "centralPackageVersionOverrideDisabled", msbuildMetadata.CentralPackageVersionOverrideDisabled);
             SetValueIfTrue(writer, "CentralPackageTransitivePinningEnabled", msbuildMetadata.CentralPackageTransitivePinningEnabled);
+            SetValueIfFalse(writer, "UsingMicrosoftNETSdk", msbuildMetadata.UsingMicrosoftNETSdk);
```
Wouldn't this be a bug when restoring with nuget.exe?

Nigusu-Allehu replied:
why would it be a bug?

nkolev92 replied:
The default for something unread is false, so I thought it could be a bug, but you're assuming default is true for this particular property, per my comment below. You can ignore this.

nkolev92 commented on src/NuGet.Core/NuGet.ProjectModel/ProjectRestoreMetadata.cs:
```diff
@@ -207,7 +222,9 @@ public bool Equals(ProjectRestoreMetadata other)
                    EqualityUtility.EqualsWithNullCheck(CentralPackageTransitivePinningEnabled, other.CentralPackageTransitivePinningEnabled) &&
-                   RestoreAuditProperties == other.RestoreAuditProperties;
+                   RestoreAuditProperties == other.RestoreAuditProperties &&
+                   EqualityUtility.EqualsWithNullCheck(UsingMicrosoftNETSdk, other.UsingMicrosoftNETSdk) &&
```
It's not a nullable bool, just use regular `==`

nkolev92 commented on src/NuGet.Core/NuGet.ProjectModel/Utf8JsonStreamReader.cs:
```diff
@@ -220,6 +221,25 @@ internal bool ReadNextTokenAsBoolOrFalse()
             return false;
         }
 
+        internal bool ReadNextTokenAsBoolOrTrue(string propertyName)
```
I now see that write if false wouldn't be a bug. 
Are we trying to save on package spec size here?

zivkan replied:
that was my suggestion, yes. Most projects now (i'm guessing, I didn't look at telemetry) are SDK style, so we can avoid CPU cycles parsing an additional property and its value when the majority of the time it's going to have that value.

nkolev92 replied:
it's not a huge perf impact and it's the type of code that can easily lead to bugs. 

It's ok if I'm the only person with the concern though.

Nigusu-Allehu committed:
address comments

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs:
```diff
@@ -511,6 +511,34 @@ private async Task<PackageSpec> GetPackageSpecAsync(ISettings settings)
             string centralPackageTransitivePinningEnabled = GetPropertySafe(_vsProjectAdapter.BuildProperties, ProjectBuildProperties.CentralPackageTransitivePinningEnabled);
             // Do not add new properties here. Use BuildProperties.GetPropertyValue instead, without DTE fallback.
 #pragma warning restore CS0618 // Type or member is obsolete
+            string skdAnalysisLevelString = _vsProjectAdapter.BuildProperties.GetPropertyValue(ProjectBuildProperties.SdkAnalysisLevel);
```
We're missing tests for the legacy code.

nkolev92 replied:
I think the review was rerequested before my latest review, so just adding this comment to make this apparent and make the nice little rerequest icon appear again :D

Nigusu-Allehu replied:
I updated the code to load the property values using `ProjectRestoreMetadata.GetSdkAnalysisLevel` which has been unit tested

nkolev92 replied:
I was referring to a LegacyPackageReference package spec creation test.

Something similar to https://github.com/NuGet/NuGet.Client/blob/4a1faa2bcc771862a0ee1d241ebbfa1c196630ab/test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/ProjectSystems/LegacyPackageReferenceProjectTests.cs#L1587-L1622.

nkolev92 commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/VSNominationUtilities.cs:
```diff
@@ -712,5 +712,44 @@ private static string[] HandleClear(string[] input)
             return input;
         }
+
+        internal static NuGetVersion? GetSdkAnalysisLevel(IReadOnlyList<IVsTargetFrameworkInfo4> targetFrameworks)
```
We're missing tests for the nomination side.

nkolev92 reviewed (commented):
We're missing some tests in some codepaths.

nkolev92 reviewed (commented):
I think we had a race condition where the review was re-requested while I was reviewing it again. Adding another comment to make sure you see the new feedback.

Nigusu-Allehu committed:
extract GetSdkAnalysisLevel

Nigusu-Allehu committed:
cleanup

Nigusu-Allehu committed:
test

Nigusu-Allehu committed:
test nominations

Nigusu-Allehu committed:
Legacy projects cleanup

Nigusu-Allehu commented:
@nkolev92 @zivkan Thanks for the reviews! please feel free to rereview. I believe I have addressed all the non-nit comments. Please let me know if there is any unaddressed comments. Thanks!

Nigusu-Allehu committed:
formatting

Nigusu-Allehu committed:
whitespace

nkolev92 commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/VSNominationUtilities.cs:
```diff
@@ -712,5 +712,44 @@ private static string[] HandleClear(string[] input)
+                }
+            }
+
+            return true;
```
It feels unintuitive to have a true as the default value of something such as UsingNetSDK. 

I think this isn't related to the roundtripping we do for the package spec, so I think false is the right thing here.

nkolev92 commented on src/NuGet.Core/NuGet.ProjectModel/ProjectRestoreMetadata.cs:
```diff
@@ -231,6 +249,40 @@ public virtual ProjectRestoreMetadata Clone()
             return clone;
         }
 
+        public static NuGetVersion GetSdkAnalysisLevel(string sdkAnalysisLevel)
```
This isn't used everywhere where it could be (example, not used in the nomination codepath).

It does promote code reuse, but we should also make sure the logic that we're adding is really worth that.

That being said, we do have a precedent, which is putting the helper methods in https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs.

zivkan replied:
I don't know how I feel about the public API either. On one hand we have a bunch of helper functions in MSBuildRestoreUtility. On the other hand, this method is used to parse the value for this data structure, so keeping the method in the same assembly as the place it's intended to be used seems like a better public API design that putting the method elsewhere in a generic "lets dump lots of random public APIs here".

I think I'm leaning towards the public API makes more sense here, if we ignore NuGet.Client precedent. There are some NuGet.Client patterns that should be stopped and modernised, but I haven't yet formed an opinion if these restore helpers are included or not.

If we keep the public API in this assembly, I'm also undecided if I'd prefer the API in this class or another. There's value in DTOs with zero business logic. But I can't think of a better class name to host the API. I really don't like `ProjectRestoreMetadataUtility` as a name.

nkolev92 reviewed (commented):
Thanks for the updates.
A few little things, and ready to 🚢

zivkan commented on src/NuGet.Core/NuGet.ProjectModel/ProjectRestoreMetadata.cs:
```diff
@@ -132,6 +134,18 @@ public class ProjectRestoreMetadata : IEquatable<ProjectRestoreMetadata>
         public RestoreAuditProperties RestoreAuditProperties { get; set; }
 
+        /// <summary>
+        /// A unified flag to help users manage their SDK warning levels. Example: 9.0.100
```
```suggestion
        /// A unified flag to help users manage their SDK warning levels. Example: 9.0.100.
```

If you hover your mouse over the property, Visual Studio will show you the IntelliSense info for the property, and will show you how the XML comment renders.

zivkan commented on src/NuGet.Core/NuGet.ProjectModel/Utf8JsonStreamReader.cs:
```diff
@@ -220,6 +222,25 @@ internal bool ReadNextTokenAsBoolOrFalse()
+            }
+            else
+            {
+                throw new ArgumentException(
```
This `ReadNextTokenAsBoolOrTrue` API is in a similar position to Nikolche's PR that we discussed in the PR review meeting today. Should we generally fail fast on invalid values, or ignore invalid values?

My larger concern is that there's already a `ReadNextTokenAsBoolOrFalse`, and it's very surprising that the two methods with almost identical names have radically different behaviour, as that one does not throw.

I expect I'm the one who proposed we should throw if the input is invalid, but this inconsistent API design is a bigger issue, so we should standardise on one or the other.

Nigusu-Allehu replied:
How about rename it to `ReadNextTokenAsBoolOrThrowAnException`?

Nigusu-Allehu committed:
Address comments

Nigusu-Allehu committed:
cleanup

Nigusu-Allehu committed:
cleanup

jeffkl commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/VSNominationUtilities.cs:
```diff
@@ -712,5 +712,30 @@ private static string[] HandleClear(string[] input)
+
+        internal static NuGetVersion? GetSdkAnalysisLevel(IReadOnlyList<IVsTargetFrameworkInfo4> targetFrameworks)
+        {
+            string? skdAnalysisLevelString = GetSingleNonEvaluatedPropertyOrNull(targetFrameworks, nameof(ProjectBuildProperties.SdkAnalysisLevel), v => v) ?? string.Empty;
```
This call to `GetSingleNonEvaluatedPropertyOrNull()` returns a `string?` so there's no need for the `?? string.Empty` right?  Then the check on line 721 is never `null` and we're calling `MSBuildRestoreUtility.GetSdkAnalysisLevel()` with `string.Empty`.

I also think its a better pattern to exit early rather than have logic in a nested `if` so the code in this method could look like this instead:
```c#
internal static NuGetVersion? GetSdkAnalysisLevel(IReadOnlyList<IVsTargetFrameworkInfo4> targetFrameworks)
{
    string? value = GetSingleNonEvaluatedPropertyOrNull(targetFrameworks, nameof(ProjectBuildProperties.SdkAnalysisLevel), v => v);
    
    if (value is null)
    {
        return null;
    }
    
    return MSBuildRestoreUtility.GetSdkAnalysisLevel(value);
}
```

Or you can simplify even more:

```C#
internal static NuGetVersion? GetSdkAnalysisLevel(IReadOnlyList<IVsTargetFrameworkInfo4> targetFrameworks)
{
    string? value = GetSingleNonEvaluatedPropertyOrNull(targetFrameworks, nameof(ProjectBuildProperties.SdkAnalysisLevel), v => v);
    
    return value is null
            ? null
            : MSBuildRestoreUtility.GetSdkAnalysisLevel(value);
}
```

The `GetUsingMicrosoftNETSdk()` is using a similar pattern but is checking for `not null`.  Reversing it would make it more readable in my opinion.

jeffkl commented on src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs:
```diff
@@ -892,7 +892,9 @@ private PackageSpec GetPackageSpec(IMSBuildProject project, IReadOnlyDictionary<
-
+            restoreMetadata.UsingMicrosoftNETSdk = MSBuildRestoreUtility.GetUsingMicrosoftNETSdk(project.GetProperty("UsingMicrosoftNETSdk"));
+            NuGetVersion skdAnalysisLevel = MSBuildRestoreUtility.GetSdkAnalysisLevel(project.GetProperty("SdkAnalysisLevel"));
+            restoreMetadata.SdkAnalysisLevel = skdAnalysisLevel ?? restoreMetadata.SdkAnalysisLevel;
```
This line confuses me.  If `skdAnalysisLevel` is `null`, then `restoreMetadata.SdkAnalysisLevel` is set to itself?  Is that needed?

Nigusu-Allehu replied:
it was supposed to be 
```CS
if (sdkAnalysisLevel is not null)
    {
        restoreMetadata.SdkAnalysisLevel = sdkAnalysisLevel;
    }
```
which would only update the value of ` restoreMetadata.SdkAnalysisLevel` if `sdkAnalysisLevel` is not null. I wanted to have this in a shorter form that's why updated it to `restoreMetadata.SdkAnalysisLevel = skdAnalysisLevel ?? restoreMetadata.SdkAnalysisLevel;` which updates the value if `sdkAnalysisLevel` is not null otherwise leaves it as is.  Another way to go about it is to not do a null check and just assign the value: `restoreMetadata.SdkAnalysisLevel = skdAnalysisLevel`. However, in the case `sdkAnalysisLevel` is null, if `restoreMetadata.SdkAnlysisLevel` has a value, we would be overwriting it. I can't think of a scenario where this happens, but I was just being extra cautious.

jeffkl replied:
In my opinion, the `??` pattern is great if you have a logical default.  In this case, you really just don't want to set the property if a value was not given so I would just go with the expanded version using the `if` statement.

jeffkl commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs:
```diff
@@ -308,6 +308,8 @@ public static PackageSpec GetPackageSpec(IEnumerable<IMSBuildItem> items)
                 result.RestoreMetadata.CentralPackageFloatingVersionsEnabled = isCentralPackageFloatingVersionsEnabled;
                 result.RestoreMetadata.CentralPackageTransitivePinningEnabled = isCentralPackageTransitivePinningEnabled;
+                result.RestoreMetadata.UsingMicrosoftNETSdk = MSBuildRestoreUtility.GetUsingMicrosoftNETSdk(specItem.GetProperty("UsingMicrosoftNETSdk"));
+                result.RestoreMetadata.SdkAnalysisLevel = MSBuildRestoreUtility.GetSdkAnalysisLevel(specItem.GetProperty("SdkAnalysisLevel")) ?? result.RestoreMetadata.SdkAnalysisLevel;
```
Same question but wanted to call it out, just answer the above one.

jeffkl commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs:
```diff
@@ -941,6 +943,40 @@ public static RestoreAuditProperties GetRestoreAuditProperties(IMSBuildItem spec
+            }
+
+            return null;
+        }
```
I find this to be a better pattern where the method exits early and has less nesting:

```suggestion
        public static NuGetVersion GetSdkAnalysisLevel(string sdkAnalysisLevel)
        {
            if (string.IsNullOrEmpty(sdkAnalysisLevel))
            {
                return null;
            }
            
            if (!NuGetVersion.TryParse(sdkAnalysisLevel, out NuGetVersion version))
            {
                throw new ArgumentException(string.Format(CultureInfo.CurrentCulture, Strings.Invalid_AttributeValue, "SdkAnalysisLevel", sdkAnalysisLevel, "9.0.100"));
            }
            
            return version;
        }
```

jeffkl commented on src/NuGet.Core/NuGet.ProjectModel/ProjectRestoreMetadata.cs:
```diff
@@ -132,6 +134,18 @@ public class ProjectRestoreMetadata : IEquatable<ProjectRestoreMetadata>
+        /// <summary>
+        /// Indicates that Microsoft.NET.Sdk is being used.
+        /// </summary>
+        public bool UsingMicrosoftNETSdk { get; set; } = true;
```
Is it dangerous for this to default to `true`?  If a code path forgets to set it based on a user supplied value, we might not catch it.  Maybe this should be a `bool?` even, so we know if it was ever set.

Nigusu-Allehu replied:
More on this discussion https://github.com/NuGet/NuGet.Client/pull/5833#discussion_r1639684059

nkolev92 replied:
Isn't that conversation more about the implied default when constructing the package spec?

Like this seems unnecessary. Failing to set this is probably a bug. I'm wondering which default makes it more apparent to us there's a bug.

nkolev92 replied:
I still think Jeff's point stands. 

We always set this, so I feel like we shouldn't go out of our way for a default. 

We can address this in a future PR though, we've held up this one long enough.

jeffkl commented on src/NuGet.Core/NuGet.ProjectModel/Strings.resx:
```diff
@@ -127,6 +127,12 @@
     <value>The pack options package type must be a string or array of strings in '{0}'.</value>
     <comment>{0} is the project.json file name.</comment>
   </data>
+  <data name="Invalid_AttributeValue" xml:space="preserve">
```
As I scroll this pull request, I've seen this same embedded resource a few times.  Do we have any shared resource we can add this to?

Nigusu-Allehu replied:
Yes there was four and I have removed two but there is two left. One in NuGet.ProjectModel and another one in  NuGet.Commands. Both of these are internal classes. I did not want to add InternalsVisibileTo and create unnecessary dependency between the two.

jeffkl commented on test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/ProjectSystems/LegacyPackageReferenceProjectTests.cs:
```diff
@@ -1621,6 +1621,171 @@ public async Task GetPackageSpec_WithNuGetAuditSuppress()
+            var actualRestoreSpec = packageSpecs.Single();
+            SpecValidationUtility.ValidateProjectSpec(actualRestoreSpec);
+
+            Assert.Equal(expectedVersion, actualRestoreSpec.RestoreMetadata.SdkAnalysisLevel);
```
Other tests in this class are using FluentAssertions so I think all new tests should use it

```suggestion
            actualRestoreSpec.RestoreMetadata.SdkAnalysisLevel.Should().Be(expectedVersion);
```

jeffkl commented on test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/ProjectSystems/LegacyPackageReferenceProjectTests.cs:
```diff
@@ -1621,6 +1621,171 @@ public async Task GetPackageSpec_WithNuGetAuditSuppress()
+            await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+            // Arrange
+            string usingSdk = "FalSe";
```
Extra points for checking case sensitivity!

Nigusu-Allehu committed:
address comments

Nigusu-Allehu committed:
cleanup

Nigusu-Allehu committed:
;

Nigusu-Allehu committed:
remove unnecessary resources

nkolev92 commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/VSNominationUtilities.cs:
```diff
@@ -21,6 +21,7 @@
 using NuGet.Shared;
 using NuGet.Versioning;
 using NuGet.VisualStudio;
+using static Microsoft.TeamFoundation.Client.CommandLine.Options;
```
I'm guessing this was accidentally added.

nkolev92 commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs:
```diff
@@ -945,36 +945,32 @@ public static RestoreAuditProperties GetRestoreAuditProperties(IMSBuildItem spec
         public static bool GetUsingMicrosoftNETSdk(string usingMicrosoftNETSdk)
         {
-            if (!string.IsNullOrEmpty(usingMicrosoftNETSdk))
+            if (string.IsNullOrEmpty(usingMicrosoftNETSdk))
```
Given that nothing is default to false (correctly), I think the smae thing needs to be done to ProjectRestoreMetadata

Nigusu-Allehu replied:
I hace removed the default true value. Which means the default is false.

nkolev92 commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs:
```diff
@@ -308,6 +308,13 @@ public static PackageSpec GetPackageSpec(IEnumerable<IMSBuildItem> items)
                 result.RestoreMetadata.CentralPackageFloatingVersionsEnabled = isCentralPackageFloatingVersionsEnabled;
                 result.RestoreMetadata.CentralPackageTransitivePinningEnabled = isCentralPackageTransitivePinningEnabled;
+                result.RestoreMetadata.UsingMicrosoftNETSdk = MSBuildRestoreUtility.GetUsingMicrosoftNETSdk(specItem.GetProperty("UsingMicrosoftNETSdk"));
+                NuGetVersion sdkAnalysisLevel = MSBuildRestoreUtility.GetSdkAnalysisLevel(specItem.GetProperty("SdkAnalysisLevel")) ?? result.RestoreMetadata.SdkAnalysisLevel;
```
Unless I'm missing something, I think line 312 and 316 should be simplified to

```cs
result.RestoreMetadata.SdkAnalysisLevel = MSBuildRestoreUtility.GetSdkAnalysisLevel(specItem.GetProperty("SdkAnalysisLevel"));
```

nkolev92 commented on test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/ProjectSystems/LegacyPackageReferenceProjectTests.cs:
```diff
@@ -1621,6 +1621,170 @@ public async Task GetPackageSpec_WithNuGetAuditSuppress()
+            actualRestoreSpec.RestoreMetadata.SdkAnalysisLevel.Should().Be(expectedVersion);
+        }
+
+        [Fact]
```
I think you can use a theory for this and the next method.

nkolev92 reviewed (commented):
A few minor comments.

Nigusu-Allehu committed:
address comments

Nigusu-Allehu committed:
Merge branch 'dev' into dev-nyenework-sdk-analysis-level

Nigusu-Allehu committed:
write if true only

Nigusu-Allehu committed:
default false everywhere

Nigusu-Allehu committed:
Fix tests

nkolev92 commented on src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs:
```diff
@@ -953,7 +953,7 @@ private static void ReadMSBuildMetadata(JsonTextReader jsonReader, PackageSpec p
             RestoreAuditProperties auditProperties = null;
             bool useMacros = MSBuildStringUtility.IsTrue(environmentVariableReader.GetEnvironmentVariable(MacroStringsUtility.NUGET_ENABLE_EXPERIMENTAL_MACROS));
             var userSettingsDirectory = NuGetEnvironment.GetFolderPath(NuGetFolderPath.UserSettingsDirectory);
-            bool usingMicrosoftNetSdk = true;
```
Just a quick clarification. 

I wasn't saying the reading for this property shouldn't default to true, I was merely saying the type shouldn't have it. 

So you can keep this. I just wouldn't initialize the type with true, because it's always something we set.

Nigusu-Allehu replied:
We have now set the default value of the property to false. On the other hand, the package spec writer updates the dgspec file only when the `usingMicrosoftNetSdk` value is false. Consequently, tests creating a package spec for other purposes encounter errors because their expected JSON lacks `usingMicrosoftNetSdk`. To address this, I modified the writer to only update if the property is true or has been manually changed from the default false to true. Similarly, I updated the reader to adopt this pattern: default to false and only update if there has been a change in the package spec JSON. This approach makes the code intuitive and eliminates the need to update tests that utilize packagespec. Essentially, while we previously defaulted to true, we now default to false.

zivkan replied:
My feedback may be idealistic or non-pragmatic, but I don't think that "a lot of tests need to be changed" is a good reason to make "worse" product changes. If tests are hard to update, then maybe the tests have tech debt that can be improved. or maybe this feature just naturally has a large impact and the work needs to be done. To me, what's more important is if the product change is good, tests aside, for the customer.

Going back to the original intent, most projects are now SDK style, and non-SDK style projects are no longer getting significant improvements, so over time we expect a larger percentage of projects to be SDK style. Maybe one day it'll reach 100%, but it'll never be 0%. The perf impact of avoiding writing `usingMicrosoftNetSdk` to the PackageSpec and the perf impact of having to read & parse it might be small, but restore is our most critical perf path, and some customers have 1000+ projects in a single restore.   I'm not saying we must micro-optimize these tiny perf improvements, but I am saying we must consider them.

Code [like this](https://github.com/dotnet/aspnetcore/blob/main/src/Shared/HttpSys/RequestProcessing/RequestHeaders.Generated.tt) isn't written because it's easy to maintain, but because the perf requirements demand it. So the question to the team is if the reduction in maintainability of having "no value in DGSpec means this value is true" is worth the perf gain. I have an opinion, but I'm just one person in the team.  But I don't think that the tests needing a lot of effort to fix is a good justification, it should be framed as customer experience, and engineering effort applies if the customer impact is the same.

Nigusu-Allehu replied:
I agree with both of you. I think we should prioritize performance. I will revert the code.

zivkan commented on src/NuGet.Core/NuGet.ProjectModel/PackageSpecWriter.cs:
```diff
@@ -210,6 +214,7 @@ private static void WriteMetadataBooleans(IObjectWriter writer, ProjectRestoreMe
             SetValueIfTrue(writer, "centralPackageFloatingVersionsEnabled", msbuildMetadata.CentralPackageFloatingVersionsEnabled);
             SetValueIfTrue(writer, "centralPackageVersionOverrideDisabled", msbuildMetadata.CentralPackageVersionOverrideDisabled);
             SetValueIfTrue(writer, "CentralPackageTransitivePinningEnabled", msbuildMetadata.CentralPackageTransitivePinningEnabled);
+            SetValueIfTrue(writer, "UsingMicrosoftNETSdk", msbuildMetadata.UsingMicrosoftNETSdk);
```
From my interpretation of https://github.com/NuGet/NuGet.Client/pull/5833#discussion_r1672759592
it was not necessary to change this to write only when true.

nkolev92 replied:
Andy's right. 
We werent' asking for this to be changed.

zivkan commented on src/NuGet.Core/NuGet.ProjectModel/Utf8JsonStreamReader.cs:
```diff
@@ -220,6 +222,25 @@ internal bool ReadNextTokenAsBoolOrFalse()
             return false;
         }
 
+        internal bool ReadNextTokenAsBoolOrTrue(byte[] propertyName)
```
I think your latest changes removed all usage of this method, in which case I think you should delete this new method.

Nigusu-Allehu replied:
We still use it in [JsonPackageSpecReader.Utf8JsonStreamReader.cs](https://github.com/NuGet/NuGet.Client/pull/5833/files#diff-5e8ef6d8ff5b223efadc3ad1d0162c8e43a663827bf73f73af89e03ffab9d2a1). I think it should be renamed to a more appropriate name: `ReadNextTokenAsBoolOrThrowAnException`

zivkan reviewed (dismissed):
I still think there's scope for improvement, but nothing blocking.

Nigusu-Allehu committed:
Rename method

Nigusu-Allehu committed:
write if false only
